File: simulator.py
class Channel:
    packet_volume = 0.020

    def __init__(self, speed):
        self.speed = speed  # kB/s

    def get_transfer_time(self, data_volume):
        return (data_volume + self.packet_volume) / self.speed


class PageAccess:
    def __init__(self, access_time, access_type, page_num):
        self.access_time = access_time
        self.access_type = access_type
        self.page_num = page_num


class AccessType:
    READ = 'read'
    WRITE = 'write'


class Simulator:
    def __init__(self, page_cnt, rw_access_history, channel):
        self.page_cnt = page_cnt
        self.rw_access_history = rw_access_history
        self.channel = channel
        self.pages_to_transfer = [i for i in range(page_cnt)]
        self.vm_location = 0
        self.cur_time = 0
        self.page_size = 4
        self.page_num_size = 0.004
        self.memory_delay = 0
        self.downtime = 0

    def _delay_memory_access(self, value):
        for i in range(len(self.rw_access_history)):
            self.rw_access_history[i].access_time += value

    def _add_memory_delay(self, value):
        self._delay_memory_access(value)
        self.memory_delay += value

    def _add_downtime(self, value):
        self._delay_memory_access(value)
        self.downtime += value

    def _get_accessed_pages(self, access_type):
        while len(self.rw_access_history) > 0 and \
                self.rw_access_history[0].access_time + \
                self.channel.get_transfer_time(self.page_num_size) < self.cur_time:
            cur_access = self.rw_access_history[0]
            self.rw_access_history.pop(0)
            if cur_access.access_type == access_type:
                yield cur_access

    def start(self):
        while len(self.pages_to_transfer) > 0:
            while len(self.pages_to_transfer) > 0:
                # page transfer
                self.cur_time += self.channel.get_transfer_time(self.page_size)
                self.pages_to_transfer.pop(0)

                # checking for page faults
                # TODO only for postcopy
                if self.vm_location == 1:
                    for access in self._get_accessed_pages(AccessType.READ):
                        if access.page_num in self.pages_to_transfer != -1:
                            # resolve page fault
                            cur_delay = self.channel.get_transfer_time(self.page_size)
                            self._add_memory_delay(cur_delay)
                            self.cur_time += cur_delay
                            self.pages_to_transfer.pop(self.pages_to_transfer.index(access.page_num))

            # update pages_to_transfer
            for access in self._get_accessed_pages(AccessType.WRITE):
                self.pages_to_transfer.append(access.page_num)

            # check whether we will be able to transfer all remaining data in less than 30 ms
            # TODO only for precopy
            cur_expected_downtime = self.channel.get_transfer_time(self.pages_to_transfer * self.page_size)
            if cur_expected_downtime < 0.03:
                self._add_downtime(cur_expected_downtime)
                self.pages_to_transfer.clear()

        return self.cur_time, self.memory_delay, self.downtime

File: simulator.cpp
#include "segment_tree.h"
#include "simulator.h"

#include <algorithm>
#include <unordered_set>

Channel::Channel(double speed, double delay) : speed_(speed), delay_(delay) {}

double Channel::get_transfer_time(double data_volume) const {
    return delay_ + (data_volume / 1000) / speed_;
}

Simulator::Simulator(size_t total_page_count, std::vector<PageAccess> access_history, Channel channel)
        : total_page_count_(total_page_count), access_history_(move(access_history)), channel_(channel) {

}

Criterias Simulator::RunMigration(MigrationScheme migration_scheme, bool optimization_flag) {
    Clear();
    if (migration_scheme == MigrationScheme::PreCopy) {
        return RunPreCopyMigration(optimization_flag);
    } else if (migration_scheme == MigrationScheme::PostCopy) {
        return RunPostCopyMigration(optimization_flag);
    } else {
        throw std::runtime_error("Unknown migration scheme");
    }
}

Criterias Simulator::RunPreCopyMigration(bool optimization_flag) {
    double transmitted_data = 0;
    int iteration_number = 0;
    std::vector<int> change_number(total_page_count_);
    while (iteration_number < PreCopyStopParameters::max_iteration_count
           && 1.0 * pages_to_transfer_.size() / total_page_count_
              > PreCopyStopParameters::ok_leave_pages_part) {
        // start iteration

        if (optimization_flag) {
            std::sort(pages_to_transfer_.begin(), pages_to_transfer_.end(),
                 [&change_number] (const int lhs, const int rhs) {
                     return change_number[lhs] < change_number[rhs];
                 });
        }

        std::unordered_set<int> next_pages_to_transfer;
        std::unordered_set<int> already_sent_pages;

        while (!pages_to_transfer_.empty()) {
            // transfer next page

            int cur_page_to_transfer = pages_to_transfer_.front();
            pages_to_transfer_.pop_front();

            double data_volume = page_size_;
            transmitted_data += data_volume;
            cur_time_ += channel_.get_transfer_time(data_volume);
            already_sent_pages.insert(cur_page_to_transfer);

            // if we have access to a page after transferring, add it to next iteration
            while (auto accessed_page_number_opt = get_next_accessed_page_number(Operation::Write)) {
                int accessed_page_number = *accessed_page_number_opt;
                if (already_sent_pages.find(accessed_page_number) != already_sent_pages.end()
                        && next_pages_to_transfer.find(accessed_page_number) == next_pages_to_transfer.end()) {
                    next_pages_to_transfer.insert(accessed_page_number);
                }
            }
        }

        pages_to_transfer_ = std::deque<int>(next_pages_to_transfer.begin(), next_pages_to_transfer.end());
    }

    double data_volume = pages_to_transfer_.size() * page_size_;
    pages_to_transfer_.clear();
    transmitted_data += data_volume;
    double downtime = channel_.get_transfer_time(data_volume);
    cur_time_ += downtime;

    return Criterias{
            .downtime = downtime,
            .delays = 0,
            .total_migration_time = cur_time_,
            .transmitted_data = transmitted_data,
            .eviction_time = cur_time_
    };
}

IndexSegment get_index_segment_by_k(int page_number, double k) {
    return {size_t((1 - k) * page_number), size_t((1 + k) * page_number)};
}

Criterias Simulator::RunPostCopyMigration(bool optimization_flag) {
    double transmitted_data = 0;
    int miss_count = 0;
    std::unordered_set<int> already_sent_pages;

    std::optional<MaxSegmentTreeOnValueIndex> segment_tree_opt;
    if (optimization_flag) {
        *segment_tree_opt = MaxSegmentTreeOnValueIndex(total_page_count_);
    }
    while (already_sent_pages.size() != total_page_count_) {
        int cur_page_to_transfer;
        if (optimization_flag) {
            cur_page_to_transfer =
                    segment_tree_opt->ComputeMax({0, total_page_count_ - 1}).index;
            segment_tree_opt->AssignMinValue(cur_page_to_transfer);
        } else {
            cur_page_to_transfer = pages_to_transfer_.front();
            pages_to_transfer_.pop_front();
        }

        double data_volume = page_size_;
        transmitted_data += data_volume;
        cur_time_ += channel_.get_transfer_time(data_volume);
        already_sent_pages.insert(cur_page_to_transfer);

        while (auto accessed_page_number_opt = get_next_accessed_page_number(Operation::Read)) {
            int accessed_page_number = *accessed_page_number_opt;
            if (already_sent_pages.find(accessed_page_number) == already_sent_pages.end()) {
                miss_count += 1;

                delays_ += channel_.get_transfer_time(page_num_size_) + channel_.get_transfer_time(page_size_);
                transmitted_data += page_num_size_ + page_size_;
                cur_time_ += channel_.get_transfer_time(page_size_);
                already_sent_pages.insert(accessed_page_number);

                if (optimization_flag) {
                    segment_tree_opt->BulkAdd(
                            get_index_segment_by_k(accessed_page_number, PostCopyOptimizationParameters::seq_k),
                            miss_count);
                    segment_tree_opt->BulkAdd(
                            get_index_segment_by_k(accessed_page_number, PostCopyOptimizationParameters::local_k),
                            miss_count);
                }
            }
        }
    }

    return Criterias{
            .downtime = 0,
            .delays = delays_,
            .total_migration_time = cur_time_,
            .transmitted_data = transmitted_data,
            .eviction_time = cur_time_,
    };
}

std::optional<int> Simulator::get_next_accessed_page_number(Operation operation) {
    while (access_history_pos_ < access_history_.size()
           && access_history_pos_ * access_gap <= cur_time_) {
        auto cur_access = access_history_[access_history_pos_++];
        if (cur_access.operation == operation) {
            return cur_access.page_number;
        }
    }
    return std::nullopt;
}

void Simulator::Clear() {
    pages_to_transfer_.clear();
    for (int i = 0; i < total_page_count_; ++i) {
        pages_to_transfer_.push_back(i);
    }
    cur_time_ = 0;
    delays_ = 0;
}


